File: packages/imanikurd/src/seerah.ts
import { loadData, loadDataAsync } from "./loadData.js";

export interface SeerahEntry {
  id: number;
  periodId: number;
  periodName: string;
  title: string;
  titleArabic: string;
  date: string;
  hijriDate: string;
  content: string; // mapped from description
  importance?: string;
}

interface RawPeriod {
  id: number;
  period: string;
  periodArabic: string;
  events: Array<{
    id: number;
    title: string;
    titleArabic: string;
    date: string;
    hijriDate: string;
    description: string;
    importance?: string;
  }>;
}

let seerahCache: SeerahEntry[] | null = null;

function flattenPeriods(periods: RawPeriod[]): SeerahEntry[] {
  const list: SeerahEntry[] = [];
  for (const p of periods) {
    if (!p.events) continue;
    for (const e of p.events) {
      list.push({
        id: e.id,
        periodId: p.id,
        periodName: p.period,
        title: e.title,
        titleArabic: e.titleArabic,
        date: e.date,
        hijriDate: e.hijriDate,
        content: e.description || "",
        importance: e.importance,
      });
    }
  }
  return list;
}

function getSeerahData(): SeerahEntry[] {
  if (!seerahCache) {
    const raw = loadData<RawPeriod[]>("seerah.json");
    seerahCache = flattenPeriods(raw);
  }
  return seerahCache;
}

async function getSeerahDataAsync(): Promise<SeerahEntry[]> {
  if (!seerahCache) {
    const raw = await loadDataAsync<RawPeriod[]>("seerah.json");
    seerahCache = flattenPeriods(raw);
  }
  return seerahCache;
}

/** Get Seerah (Prophet Muhammad biography) entries */
export function getSeerah(): SeerahEntry[] {
  return getSeerahData();
}

/** Get Seerah (Prophet Muhammad biography) entries asynchronously */
export async function getSeerahAsync(): Promise<SeerahEntry[]> {
  return getSeerahDataAsync();
}

/** Get a single Seerah entry by id */
export function getSeerahEntry(id: number): SeerahEntry | undefined {
  return getSeerahData().find((s) => s.id === id);
}

/** Get a single Seerah entry by id asynchronously */
export async function getSeerahEntryAsync(id: number): Promise<SeerahEntry | undefined> {
  const data = await getSeerahDataAsync();
  return data.find((s) => s.id === id);
}

/** Search Seerah content */
export function searchSeerah(query: string): SeerahEntry[] {
  const normalized = query.trim().toLowerCase();
  if (!normalized) return [];

  return getSeerahData().filter(
    (s) =>
      s.title.toLowerCase().includes(normalized) ||
      s.content.toLowerCase().includes(normalized)
  );
}

/** Search Seerah content asynchronously */
export async function searchSeerahAsync(query: string): Promise<SeerahEntry[]> {
  const normalized = query.trim().toLowerCase();
  if (!normalized) return [];

  const data = await getSeerahDataAsync();
  return data.filter(
    (s) =>
      s.title.toLowerCase().includes(normalized) ||
      s.content.toLowerCase().includes(normalized)
  );
}

File: packages/imanikurd_flutter/lib/src/seerah.dart
import 'load_data.dart';
import 'models.dart';

List<SeerahEntry>? _seerahCache;

class _RawPeriod {
  final int id;
  final String period;
  final List<Map<String, dynamic>> events;

  _RawPeriod({required this.id, required this.period, required this.events});

  factory _RawPeriod.fromJson(Map<String, dynamic> json) => _RawPeriod(
        id: json['id'] as int,
        period: json['period'] as String,
        events: (json['events'] as List<dynamic>? ?? [])
            .map((e) => e as Map<String, dynamic>)
            .toList(),
      );
}

List<SeerahEntry> _flattenPeriods(List<_RawPeriod> periods) {
  final list = <SeerahEntry>[];
  for (final period in periods) {
    for (final event in period.events) {
      list.add(
        SeerahEntry(
          id: event['id'] as int,
          periodId: period.id,
          periodName: period.period,
          title: event['title'] as String,
          titleArabic: event['titleArabic'] as String,
          date: event['date'] as String,
          hijriDate: event['hijriDate'] as String,
          content: (event['description'] as String?) ?? '',
          importance: event['importance'] as String?,
        ),
      );
    }
  }
  return list;
}

Future<List<SeerahEntry>> _getSeerahData() async {
  _seerahCache ??= _flattenPeriods(
    (await loadDataAsync<List<dynamic>>('seerah.json'))
        .map((e) => _RawPeriod.fromJson(e as Map<String, dynamic>))
        .toList(),
  );
  return _seerahCache!;
}

Future<List<SeerahEntry>> getSeerah() => _getSeerahData();

Future<SeerahEntry?> getSeerahEntry(int id) async {
  final data = await _getSeerahData();
  for (final entry in data) {
    if (entry.id == id) return entry;
  }
  return null;
}

Future<List<SeerahEntry>> searchSeerah(String query) async {
  final normalized = query.trim().toLowerCase();
  if (normalized.isEmpty) return [];
  final data = await _getSeerahData();
  return data
      .where(
        (s) =>
            s.title.toLowerCase().contains(normalized) ||
            s.content.toLowerCase().contains(normalized),
      )
      .toList();
}
